### PR TITLE
kv: implement server-side negotiation fast-path and Txn.NegotiateAndSend

### LINKED ISSUE #67551: kv: introduce bounded staleness BatchRequest Header fields
Part of #66020.
Needed for **stage 1**.

See https://github.com/cockroachdb/cockroach/blob/025c53decd8eb00cb70a1d410692026c9f8614dc/docs/RFCS/20210519_bounded_staleness_reads.md#protobuf-api-changes.

### LINKED ISSUE #67552: kv: implement bounded staleness server-side negotiation fast-path
Part of #66020.
Needed for **stage 1**.

See https://github.com/cockroachdb/cockroach/blob/025c53decd8eb00cb70a1d410692026c9f8614dc/docs/RFCS/20210519_bounded_staleness_reads.md#server-side-negotiation-fast-path.

### LINKED ISSUE #67553: kv: implement Txn.NegotiateAndSend
Part of #66020.
Needed for **stage 1**.

See https://github.com/cockroachdb/cockroach/blob/025c53decd8eb00cb70a1d410692026c9f8614dc/docs/RFCS/20210519_bounded_staleness_reads.md#implementation-of-txnnegotiateandsend.

### PR DESCRIPTION
**Pull request by nvb**
Fixes #67551.
Fixes #67552.
Fixes #67553.
Touches #67562.

Bounded-staleness read orchestration consists of two phases - negotiation and execution. Negotiation determines the timestamp to run the query at in order to ensure that the read will not block on replication or on conflicting transactions. Execution then uses this timestamp to run the read request.

This commit implements the bounded staleness server-side negotiation fast-path. This fast-path allows a bounded staleness read request that lands on a single range to perform its negotiation phase and execution phase in a single RPC.

The server-side negotiation fast-path provides two benefits:
1. it avoids two network hops in the common-case where a bounded staleness read is targeting a single range. This in an important performance optimization for single-row point lookups.
2. it provides stronger guarantees around minimizing staleness during bounded staleness reads. Bounded staleness reads that hit the server-side fast-path use their target replica's most up-to-date resolved timestamp, so they are as fresh as possible. Bounded staleness reads that miss the fast-path and perform explicit negotiation (see below) consult a cache, so they may use an out-of-date, suboptimal resolved timestamp, as long as it is fresh enough to satisfy the staleness bound of the request.

The commit then uses this new functionality to implement the `(*Txn).NegotiateAndSend` method detailed in the bounded staleness RFC. `NegotiateAndSend` is a specialized version of `Send` that is capable of orchestrating a bounded-staleness read through a transaction, given a read-only BatchRequest with a `min_timestamp_bound` set in its Header. If the method returns successfully, the transaction will have been given a fixed timestamp equal to the timestamp that the read-only request was evaluated at.

### CODE REVIEW

File: pkg/roachpb/errors.go [hunk 1]
```diff
@@ -1280,3 +1281,31 @@ func (e *OptimisticEvalConflictsError) Type() ErrorDetailType {
...
+	minTimestampBound, resolvedTimestamp hlc.Timestamp,
+) *MinTimestampBoundUnsatisfiableError {
+	return &MinTimestampBoundUnsatisfiableError{
+		MinTimestampBound: minTimestampBound,
+		ResolvedTimestamp: resolvedTimestamp,
+	}
+}
+
+func (e *MinTimestampBoundUnsatisfiableError) Error() string {
+	return e.message(nil)
```
Comment (otan) [inline pkg/roachpb/errors.go:1297]: should we add a pgcode to this? or is there a place up the stack we usually add a pgcode to this kind of thing?

**Review by otan — COMMENTED** [review]
looks good to me (can build on top of this), but can't verify the KV stuff as i'm not familiar enough with it all

File: pkg/roachpb/errors.go (hunk 1 above)
Comment (nvanbenschoten) [reply pkg/roachpb/errors.go:1297]: We don't have a way that I'm aware of to attach pgcodes to KV-layer errors. Instead, we have a translation layer in `pkg/sql/row`. See `ConvertBatchError` (which is growing a bit in https://github.com/cockroachdb/cockroach/pull/68042).

**Comment by nvanbenschoten** [comment]
@otan I appended a third commit here which adds a new `min_timestamp_bound_strict` flag to the `BatchRequest.Header` struct. This flag specifies whether a bounded staleness read whose `min_timestamp_bound` cannot be satisfied by the first replica it visits without blocking should be rejected with a `MinTimestampBoundUnsatisfiableError` or will be redirected to the leaseholder and permitted to block on conflicting transactions. You'll need this to implement the `nearest_only` option as well as to implement the descriptor iteration logic that steps back across table descriptor versions.

@aliher1911 I assigned you as a reviewer on this (just the last 3 commits). Let me know if you'd like to walk through this together at some point. I'm happy to explain the code or the context surrounding this change.

**Comment by nvanbenschoten** [comment]
@otan I added one more commit here with the `max_timestamp_bound` field we had talked about.

**Comment by nvanbenschoten** [comment]
@aayushshah15 thanks for offering to give this a review. It's now rebased on master, so it no longer has any dependencies.

**Review by aayushshah15 — APPROVED** [review]
:lgtm: mod that one question. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68194#-:-MhFpOFi5tl0SNO9kK2H:bguzjge)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911, @nvanbenschoten, and @otan)

___
*[pkg/kv/txn.go, line 1129 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68194#-MhFj7wG6Myb02_G67dB:-MhFj7wG6Myb02_G67dC:b-olcsjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c63a6a33026bb13c83142d658f9ef747434a117/pkg/kv/txn.go#L1129)):*
> ```Go
> 	assert(!ba.IsLocking(), "batch must not be locking")
> 	assert(txn.typ == RootTxn, "txn must be root")
> 	assert(!txn.CommitTimestampFixed(), "txn commit timestamp must not be fixed")
> ```

Would it ever be possible for the SQL layer to set the `max_timestamp_bound` lower than a given `min_timestamp_bound`? is it worth asserting against that here?
I see that `executeServerSideBoundedStalenessNegotiation` could return a timestamp lower than the `min_timestamp_bound` in such a case.
___
*[pkg/server/node.go, line 1037 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68194#-MhFiRG08hi8Msgtx1fB:-MhFiRG08hi8Msgtx1fC:bhbwa8s) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68ba74b5617f4db0a682533a4437ac8e28f7139/pkg/server/node.go#L1037)):*
> ```Go
> 		ctx, newSpan = tracing.EnsureChildSpan(ctx, tr, opName)
> 		// Set the same span.kind tag as the gRPC interceptor.
> 		newSpan.SetTag(ext.SpanKindRPCServer.Key, ext.SpanKindRPCServer.Value)
> ```

OOC, was this needed to get the `OnlyFollowerReads` helper to work?

**Review by nvb — COMMENTED** [review]
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68194#-:-MhFzs9_5dy0PUWY1jXI:b-9qdoz3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aayushshah15, @aliher1911, and @otan)

___
*[pkg/kv/txn.go, line 1129 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/68194#-MhFj7wG6Myb02_G67dB:-MhFyY6T7KRqtKIQ9gxk:bb4sx2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c63a6a33026bb13c83142d658f9ef747434a117/pkg/kv/txn.go#L1129)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

Would it ever be possible for the SQL layer to set the `max_timestamp_bound` lower than a given `min_timestamp_bound`? is it worth asserting against that here?
I see that `executeServerSideBoundedStalenessNegotiation` could return a timestamp lower than the `min_timestamp_bound` in such a case.

</blockquote></details>

Good point. I added a check for this.
___
*[pkg/server/node.go, line 1037 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/68194#-MhFiRG08hi8Msgtx1fB:-MhFyHUT70EDeeVvpegT:b-sc3kuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/f68ba74b5617f4db0a682533a4437ac8e28f7139/pkg/server/node.go#L1037)):*
<details><summary><i>Previously, aayushshah15 (Aayush Shah) wrote…</i></summary><blockquote>

OOC, was this needed to get the `OnlyFollowerReads` helper to work?
</blockquote></details>

Yes, exactly. `OnlyFollowerReads` matches on `sp.Tags["span.kind"] == "server"`, which we weren't setting on the local RPC-bypass.

**Pull request merged**